src/Functions/FunctionsStringDistance.cpp

```diff
@@ -271,6 +272,222 @@ struct ByteEditDistanceImpl
     }
 };
 
+struct ByteDamerauLevenshteinDistanceImpl
+{
+    using ResultType = UInt64;
+    static constexpr size_t max_string_size = 1u << 16;
+
+    static ResultType inline process(
+        const char * __restrict haystack, size_t haystack_size, const char * __restrict needle, size_t needle_size)
+    {
+        /// Safety threshold against DoS, since we use two array to calculate the distance.
+        if (haystack_size > max_string_size || needle_size > max_string_size)
+            throw Exception(
+                ErrorCodes::TOO_LARGE_STRING_SIZE,
+                "The string size is too big for function editDistance, "
+                "should be at most {}", max_string_size);
+
+        // short cut cases:
+  	// - null strings
+  	// - zero length strings
+  	// - identical length and value strings
+  	if (haystack == nullptr || needle == nullptr ) return -1;
+  	if (haystack_size == 0) return needle_size;
+  	if (needle_size == 0) return haystack_size;
+  	if (haystack_size == needle_size && memcmp(haystack, needle, haystack_size) == 0) return 0;
+
+  	int i;
+  	int j;
+  	int l_cost;
+  	int ptr_array_length = static_cast<int>(sizeof(int*) * (haystack_size + 1));
+  	int int_array_length = static_cast<int>(sizeof(int) * (needle_size + 1) * (haystack_size + 1));
+	int s1len = static_cast<int>(haystack_size);
+	int s2len = static_cast<int>(needle_size);
+	// Dynamically allocate memory for the 2D array
+  	// Allocating a 2D array (with d being an array of pointers to the start of the rows)
+  	int** d = reinterpret_cast<int**>(new int*[ptr_array_length]);
+  	int* rows = reinterpret_cast<int*>(new int[int_array_length]);
+  	// Setting the pointers in the pointer-array to the start of (s2len + 1) length
+  	// intervals and initializing its values based on the mentioned algorithm.
+  	
+	for (i = 0; i <= s1len; ++i) {
```

Comment:
So is this the "Optimal string alignment distance" (--> Wikipedia) or the "Distance with adjacent transpositions". Please add comments.